client-rpc/src/server.rs

```diff
@@ -35,23 +37,57 @@ impl Server {
         })
     }
 
-    pub(crate) fn start(&self) -> Result<()> {
-        let storage = SledStorage::new(&self.storage_dir)?;
-        let tendermint_client = RpcClient::new(&self.tendermint_url);
-        let signer = DefaultSigner::new(storage.clone());
-        let transaction_builder =
-            DefaultTransactionBuilder::new(signer, tendermint_client.genesis()?.fee_policy());
-        let index = DefaultIndex::new(storage.clone(), tendermint_client);
-        let wallet_client = DefaultWalletClient::builder()
-            .with_wallet(storage)
-            .with_transaction_read(index)
-            .with_transaction_write(transaction_builder)
-            .build()?;
-        let wallet_rpc = WalletRpcImpl::new(wallet_client, self.chain_id);
+    pub fn start_wallet(&self, io: &mut IoHandler, storage: SledStorage) -> Result<()> {
+        {
+            let tendermint_client = RpcClient::new(&self.tendermint_url);
+            let signer = DefaultSigner::new(storage.clone());
+            let transaction_builder =
+                DefaultTransactionBuilder::new(signer, tendermint_client.genesis()?.fee_policy());
+            let index = DefaultIndex::new(storage.clone(), tendermint_client);
+            let wallet_client = DefaultWalletClient::builder()
+                .with_wallet(storage)
+                .with_transaction_read(index)
+                .with_transaction_write(transaction_builder)
+                .build()?;
+            let wallet_rpc = WalletRpcImpl::new(wallet_client, self.chain_id);
+            io.extend_with(wallet_rpc.to_delegate());
+        }
+        Ok(())
+    }
 
+    pub fn start_client(&self, io: &mut IoHandler, storage: SledStorage) -> Result<()> {
+        {
+            let tendermint_client = RpcClient::new(&self.tendermint_url);
+            let tendermint_client2 = RpcClient::new(&self.tendermint_url);
```

Comment:
Can `tendermint_client` and `signer` be cloned ?
I guess the reason for this duplication here is that they are both needed in `DefaultNetworkOpsClient` and `DefaultWalletClient` + `DefaultNetworkOpsClient` needs `DefaultWalletClient`.
There are probably various options how to simplify it:

1. `DefaultWalletClient` exposes / allows its `signer` and `tendermint_client` to be borrowed. So, `DefaultNetworkOpsClient` only needs `wallet_client`.
2. the functionality needed by `DefaultNetworkOpsClient` is moved to those common crates, so that `DefaultNetworkOpsClient` doesn't need `DefaultWalletClient`

Reply:
i see,
borrowing from "new" makes management complicated.
how about "move" semantics instead of borrowing.

also we don't use "Rc" or "Arc" to share, it's better to use "move" instead of borrowing in "new".

Reply:
because if multiple share is necessary, borrowing, lifetime management is difficult.